src/tests/catwalk_tests/test_evaluation.py

```diff
@@ -1,22 +1,243 @@
+        "query": """\
+            select entity_id, as_of_date
+            from {results_schema}.predictions
+            where entity_id % 2 = 1
+            and as_of_date in (SELECT(UNNEST(ARRAY{as_of_dates}::timestamp[])))
+            and model_id = {model_id}
+        """,
+        "hash": "fghij",
+    },
+    {
+        "name": "empty",
+        "query": """\
+            select entity_id, as_of_date
+            from {results_schema}.predictions
+            where entity_id = -1
+            and as_of_date in (SELECT(UNNEST(ARRAY{as_of_dates}::timestamp[])))
+            and model_id = {model_id}
+        """,
+        "hash": "abcde",
+    },
+]
+
+
+def setup_results_schema(db_engine, num_entities, labels):
+    # set up data, randomly generated by the factories but conforming
+    # generally to what we expect model_metadata and results schemas data
+    # to look like
+    model = ModelFactory()
+
+    class ImmediateTestPredictionFactory(TestPredictionFactory):
+        as_of_date = factory.LazyAttribute(
+            lambda o: o.model_rel.train_end_time
+        )
+    class ImmediateTrainPredictionFactory(TrainPredictionFactory):
+        as_of_date = factory.LazyAttribute(
+            lambda o: o.model_rel.train_end_time
+        )
+    
+    for entity in range(0, num_entities):
+        ImmediateTestPredictionFactory(
+            model_rel=model,
+            entity_id=entity,
+            label_value=labels[entity],
+        )
+        ImmediateTrainPredictionFactory(
+            model_rel=model,
+            entity_id=entity,
+            label_value=labels[entity],
+        )
+
+    session.commit()
+        
+    scores = []
+    labels = []
+    for row in db_engine.execute(
+        """\
+        select label_value, score
+        from test_results.predictions
+        order by entity_id
+        """
+    ):
+        labels.append(row[0])
+        scores.append(float(row[1]))
+
+    model_metadata = [
+        row 
+        for row in db_engine.execute(
+            """\
+            select model_id, train_matrix_uuid, train_end_time
+            from model_metadata.models
+            """
+        )
+    ][0]
+
+    return scores, model_metadata[0], model_metadata[1], model_metadata[2]
+
+
+def fake_matrix_from_db(db_engine, matrix_type, model_id, num_entities, as_of_date):
+        fake_matrix_store = MockMatrixStore(
+            matrix_type,
+            model_id,
+            num_entities,
+            db_engine,
+            init_labels=pandas.read_sql(
+                "select entity_id, as_of_date, label_value from train_results.predictions",
+                db_engine
+            ).set_index(["entity_id", "as_of_date"]).label_value,
+            init_as_of_dates=[as_of_date]
+        )
+        return fake_matrix_store
+
+
+def test_all_same_labels():
+    for label_value in [0, 1]:
+        with testing.postgresql.Postgresql() as postgresql:
+            db_engine = create_engine(postgresql.url())
```

Comment:
**if you like**, (cause it's not a big deal for this feature), there's now a [handy pytest fixture](https://github.com/dssg/triage/blob/master/src/tests/conftest.py) that does this most basic setup for you:

```
def test_all_same_labels(db_engine):
    ensure_db(db_engine)
    init_engine(db_engine)
    ...
```

moreover, if you like, [you can then extend this](https://github.com/dssg/triage/blob/1585076e78dfb9d84a442509df6ff0c9e83d2565/src/tests/architect_tests/test_feature_generators.py) for this set of tests:

```
@pytest.fixture(name='test_engine', scope='function')
def fixture_test_engine(db_engine):
    """Local extension to the shared db_engine fixture to set up test
    database tables.
     """
    ensure_db(db_engine)
    init_engine(db_engine)
    # etc.
    return db_engine

def test_all_same_labels(test_engine):
    # all done!
    # (though my database engine is now locally called "test_engine"
    num_entities = 5
    ...
```

finally, you can get better support from pytest for these parameterized tests -- such that [each test parameter loop is treated as a separate test case](https://github.com/dssg/triage/blob/master/src/tests/test_experiments.py#L40) -- 

```
@pytest.mark.parameterize(
    ('label_value',),
    [
        (0,),
        (1,),
    ],
)
def test_all_same_labels(test_engine, label_value):
    # no need for "for" loop either!
    num_entities = 5
```

(and optionally assign a module-level reference to this decorator rather than define inline.)

...all just options for making these tests less repetitive, more focused on their core assertions, and (ostensibly) easier to understand.

Reply:
coooool! i really hated writing these tests, and i think this would have made it so much better.